File: Algorithms/Python/pair_product.py
from __future__ import annotations


def pair_product(arr: list[int], target: int) -> tuple[int, int]:
    if len(arr) < 2:
        return -1, -1

    seen = {}
    for i, num in enumerate(arr):
        if target % num == 0 and target // num in seen:
            return seen[target // num], i
        seen[num] = i

    return -1, -1


if __name__ == "__main__":
    print(pair_product([10, 20, 9, 40], 400))  # (1, 3)
    print(pair_product([10, 20, 9, 40], 190))  # (-1, -1)
    print(pair_product([10, 20, 9, 40], 90))  # (0, 2)

File: Algorithms/Java/PairProduct.java
import java.util.HashMap;

class PairProduct {
    public static void main(final String[] args) {
        int[] arr = { 1, 2, 3, 4, 5 };
        int x = 10;
        int[] result = pairProduct(arr, x);
        if (result[0] == -1 && result[1] == -1) {
            System.out.println("No pair found");
        } else {
            System.out.println("Pair found at index " + result[0] + " and " + result[1]);
        }
    }

    static int[] pairProduct(final int[] arr, final int target) {
        final HashMap<Integer, Integer> map = new HashMap<>();
        final int[] result = { -1, -1 };

        for (int i = 0; i < arr.length; i++) {
            if (map.containsKey(target / arr[i])) {
                result[0] = map.get(target / arr[i]);
                result[1] = i;
                return result;
            }
            map.put(arr[i], i);
        }
        return result;
    }
}
